chromium_src/third_party/blink/renderer/modules/eventsource/event_source.cc

```diff
@@ -0,0 +1,62 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "third_party/blink/renderer/modules/eventsource/event_source.h"
+
+#include "brave/third_party/blink/renderer/core/farbling/brave_session_cache.h"
+#include "third_party/blink/public/common/features.h"
+#include "third_party/blink/public/common/scheme_registry.h"
+#include "third_party/blink/public/platform/web_content_settings_client.h"
+#include "third_party/blink/renderer/core/dom/document.h"
+#include "third_party/blink/renderer/core/frame/dom_window.h"
+#include "third_party/blink/renderer/core/streams/readable_byte_stream_controller.h"
+
+#define ConnectTimerFired(...) ConnectTimerFired_ChromiumImpl(__VA_ARGS__)
+#define close() close_ChromiumImpl()
+
+#include "src/third_party/blink/renderer/modules/eventsource/event_source.cc"
+
+#undef ConnectTimerFired
+#undef close
+
+namespace blink {
+
+void EventSource::ConnectTimerFired(TimerBase* timer_base) {
+  BraveConnect();
+}
+
+void EventSource::BraveConnect() {
+  if (base::FeatureList::IsEnabled(blink::features::kRestrictEventSourcePool)) {
+    ExecutionContext* execution_context = GetExecutionContext();
+    if (blink::WebContentSettingsClient* settings =
+            brave::GetContentSettingsClientFor(execution_context)) {
+      const bool is_extension = CommonSchemeRegistry::IsExtensionScheme(
+          execution_context->GetSecurityOrigin()->Protocol().Ascii());
+      if (!is_extension &&
+          settings->GetBraveFarblingLevel() != BraveFarblingLevel::OFF) {
+        event_source_in_use_tracker_ =
+            ResourcePoolLimiter::GetInstance().IssueResourceInUseTracker(
+                execution_context,
+                ResourcePoolLimiter::ResourceType::kEventSource);
+        if (!event_source_in_use_tracker_) {
+          AbortConnectionAttempt();
+          return;
+        }
+      }
+    }
+  }
+  Connect();
+}
+
+void EventSource::close() {
+  if (base::FeatureList::IsEnabled(blink::features::kRestrictWebSocketsPool)) {
+    if (event_source_in_use_tracker_ != nullptr) {
+      event_source_in_use_tracker_.reset();
```

Comment:
pls make sure you cover all "close/invalidate"-type events to reset the tracker properly.

Reply:
Thanks for bringing this up -- I found and implemented two additional cases where the tracker should be reset.